pkg/imgpkg/cmd/push_test.go

```diff
@@ -116,26 +119,56 @@ func TestNestedImgpkgDirError(t *testing.T) {
 	}
 }
 
-func TestBundleFlagWithoutDirectoryError(t *testing.T) {
-	tempDir := os.TempDir()
-	pushDir := filepath.Join(tempDir, "imgpkg-push-units-bundle-without-dir")
-	defer Cleanup(pushDir)
-
-	// cleanup any previous state
-	Cleanup(pushDir)
-	err := os.Mkdir(pushDir, 0700)
-	if err != nil {
-		t.Fatalf("Failed to setup test: %s", err)
-	}
-
-	push := PushOptions{FileFlags: FileFlags{Files: []string{pushDir}}, BundleFlags: BundleFlags{Bundle: "foo"}}
-	err = push.Run()
-	if err == nil {
-		t.Fatalf("Expected validations to err, but did not")
+func TestBundleDirectoryErrors(t *testing.T) {
+	testCases := []struct {
+		name            string
+		expectedError   string
+		createBundleDir bool
+	}{
+		{
+			name:            "bundle but no imagesLock",
+			createBundleDir: true,
+			expectedError:   "The bundle expected .imgpkg/images.yml to exist, but it wasn't found",
+		},
+		{
+			name:            "no bundle",
+			createBundleDir: false,
+			expectedError:   "This directory is not a bundle. It it is missing .imgpkg/images.yml",
+		},
 	}
 
-	if !strings.Contains(err.Error(), "Expected one '.imgpkg' dir, got 0") {
-		t.Fatalf("Expected error to contain message about missing .imgpkg dir, got: %s", err)
+	for _, tc := range testCases {
+		f := func(t *testing.T) {
+			tempDir := os.TempDir()
+			pushDir := filepath.Join(tempDir, "imgpkg-push-dir")
+
+			err := os.Mkdir(pushDir, 0700)
+			require.NoError(t, err)
+
+			if tc.createBundleDir {
+				err = createEmptyBundleDir(pushDir)
+				if err != nil {
+					Cleanup(pushDir)
```

Comment:
Do you think that instead of having this here we could do a `defer Cleanup(pushDir)`?
That way we would be able to use the assertions from testify?

Reply:
Yes, every time I have this type of code I start out this way then remember that defers are not supposed to be used inside loops. This is because it only execute when the function returns, not the loop.

Reply:
In this case, we have a function that will be run in the `t.Run`. Does that make a difference?

Reply:
Great catch! Looks so much better now.